haystack/pipelines/base.py

```diff
@@ -501,43 +487,64 @@ def add_node(self, component, name: str, inputs: List[str]):
-            )
+            if not len(inputs) == 1 and inputs[0].split(".")[0] == self.root_node:
+                raise PipelineConfigError(
+                    f"The '{name}' node can only input from {self.root_node}. "
+                    f"Set the 'inputs' parameter to ['{self.root_node}']"
+                )
             self.graph.add_edge(self.root_node, name, label="output_1")
             return
 
-        for i in inputs:
-            if "." in i:
-                [input_node_name, input_edge_name] = i.split(".")
-                assert "output_" in input_edge_name, f"'{input_edge_name}' is not a valid edge name."
+        for input_node in inputs:
+            if "." in input_node:
+                [input_node_name, input_edge_name] = input_node.split(".")
+                if not "output_" in input_edge_name:
+                    raise PipelineConfigError(f"'{input_edge_name}' is not a valid edge name.")
+
                 outgoing_edges_input_node = self.graph.nodes[input_node_name]["component"].outgoing_edges
-                assert int(input_edge_name.split("_")[1]) <= outgoing_edges_input_node, (
-                    f"Cannot connect '{input_edge_name}' from '{input_node_name}' as it only has "
-                    f"{outgoing_edges_input_node} outgoing edge(s)."
-                )
+                if not int(input_edge_name.split("_")[1]) <= outgoing_edges_input_node:
+                    raise PipelineConfigError(
+                        f"Cannot connect '{input_edge_name}' from '{input_node_name}' as it only has "
+                        f"{outgoing_edges_input_node} outgoing edge(s)."
+                    )
             else:
-                outgoing_edges_input_node = self.graph.nodes[i]["component"].outgoing_edges
-                assert outgoing_edges_input_node == 1, (
-                    f"Adding an edge from {i} to {name} is ambiguous as {i} has {outgoing_edges_input_node} edges. "
-                    f"Please specify the output explicitly."
-                )
-                input_node_name = i
+                try:
+                    outgoing_edges_input_node = self.graph.nodes[input_node]["component"].outgoing_edges
+                except KeyError as e:
+                    raise PipelineConfigError(
+                        message=f"Cannot find node '{input_node}'. Make sure you're not using more "
+                        f"than one root node ({valid_root_nodes}) in the same pipeline and that a node "
+                        f"called '{input_node}' is defined.",
+                        source=e,
+                    )
+
+                if not outgoing_edges_input_node == 1:
+                    raise PipelineConfigError(
+                        f"Adding an edge from {input_node} to {name} is ambiguous as {input_node} has {outgoing_edges_input_node} edges. "
+                        f"Please specify the output explicitly."
+                    )
```

Comment:
Fair point :+1: I like to minimize the content of try blocks to avoid catching unintended errors. However here there's really nothing to raise, so I'll move it in.